lang/java/avro/src/main/java/org/apache/avro/file/XZCodec.java

```diff
@@ -57,62 +52,22 @@ public String getName() {
   }
 
   @Override
-  public ByteBuffer compress(ByteBuffer data) throws IOException {
-    ByteArrayOutputStream baos = getOutputBuffer(data.remaining());
-    OutputStream ios = new XZCompressorOutputStream(baos, compressionLevel);
-    writeAndClose(data, ios);
-    return ByteBuffer.wrap(baos.toByteArray());
+  protected OutputStream compressedStream(OutputStream output)
+      throws IOException {
+    return new XZCompressorOutputStream(output, compressionLevel);
   }
 
   @Override
-  public ByteBuffer decompress(ByteBuffer data) throws IOException {
-    ByteArrayOutputStream baos = getOutputBuffer(data.remaining());
-    InputStream bytesIn = new ByteArrayInputStream(
-      data.array(),
-      data.arrayOffset() + data.position(),
-      data.remaining());
-    InputStream ios = new XZCompressorInputStream(bytesIn);
-    try {
-      IOUtils.copy(ios, baos);
-    } finally {
-      ios.close();
-    }
-    return ByteBuffer.wrap(baos.toByteArray());
+  protected InputStream uncompressedStream(InputStream input)
+      throws IOException {
+    return new XZCompressorInputStream(input);
   }
 
-  private void writeAndClose(ByteBuffer data, OutputStream to) throws IOException {
-    byte[] input = data.array();
-    int offset = data.arrayOffset() + data.position();
-    int length = data.remaining();
-    try {
-      to.write(input, offset, length);
-    } finally {
-      to.close();
-    }
-  }
-
-  // get and initialize the output buffer for use.
-  private ByteArrayOutputStream getOutputBuffer(int suggestedLength) {
-    if (null == outputBuffer) {
-      outputBuffer = new ByteArrayOutputStream(suggestedLength);
-    }
-    outputBuffer.reset();
-    return outputBuffer;
-  }
-
-  @Override
-  public int hashCode() {
-    return compressionLevel;
-  }
+  @Override public int hashCode() { return getName().hashCode(); }
 
   @Override
-  public boolean equals(Object obj) {
-    if (this == obj)
-      return true;
-    if (getClass() != obj.getClass())
-      return false;
-    XZCodec other = (XZCodec)obj;
```

Comment:
The equals method for the XZ codec was wrong -- according to the specification of Codec, it should equal if it is mutually decompressible.  The compression level is used for the compressor but does not affect decompressibility.

I made the implementation of hashCode and equals consistent across the codecs when appropriate.